src/zarr/registry.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from zarr.abc.codec import Codec, CodecPipeline
+    from zarr.buffer import Buffer, NDBuffer
+
+from importlib.metadata import EntryPoint
+from importlib.metadata import entry_points as get_entry_points
+
+from zarr.config import BadConfigError, config
+
+__codec_registry: dict[str, dict[str, type[Codec]]] = {}
```

Comment:
This code could be a bit DRY-er. Maybe add a generic class (e.g. namedtuple) for `__*_registry` and `__lazy_load_*`?